lib/crystalball/map_storage/yaml_storage.rb

```diff
@@ -17,11 +17,19 @@ def clear!
       end
 
       def load
-        YAML.safe_load(path.read) if path.exist?
+        metadata, *cases = path.read.split("---\n").reject(&:empty?).map do |yaml|
+          YAML.safe_load(yaml, [Symbol])
+        end
+        cases = cases.inject(&:merge)
+
+        Object.const_get(metadata[:type]).new(self, metadata: metadata, cases: cases)
       end
 
-      def dump(map)
-        path.open('a') { |f| f.write YAML.dump(map) }
+      def dump(map, exclude_metadata: false)
+        path.open('a') do |f|
+          f.write YAML.dump(map.serialize[:metadata]) unless exclude_metadata
```

Comment:
I don't think it should be responsibility responsibility of `YAMLStorage` when to store what. Rather we should have 2 methods `dump_metadata` and `dump_cases` and let storage engines only decide how do we store the data, not when do we store the data.

Reply:
YAMLStorage doesn't decide. StandardMap decides :)